File: day032_odd_even_sum/solution.cpp
#include <bits/stdc++.h>
using namespace std;

// ─────────────────────────────────────────────────────────────────────────
// OPTIMIZED — O(n) time | O(1) space
// Single pass — add each number to odd or even accumulator
// ─────────────────────────────────────────────────────────────────────────
void oddEvenSum(vector<int>& arr) {
    long long oddSum  = 0;                       // sum of all odd numbers
    long long evenSum = 0;                       // sum of all even numbers

    for (int num : arr) {
        if (abs(num % 2) == 1) {                 // abs handles negative numbers
            oddSum  += num;                      // odd bucket
        } else {
            evenSum += num;                      // even bucket
        }
    }

    cout << oddSum  << endl;                     // print odd sum
    cout << evenSum << endl;                     // print even sum
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n;
    cin >> n;

    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    oddEvenSum(arr);
    return 0;
}

File: day032_odd_even_sum/solution.java
import java.util.Scanner;

public class solution {

    // ─────────────────────────────────────────────────────────────────────
    // OPTIMIZED — O(n) time | O(1) space
    // Single pass — add each number to odd or even accumulator
    // ─────────────────────────────────────────────────────────────────────
    static long[] oddEvenSum(int[] arr) {
        long oddSum  = 0;                        // sum of all odd numbers
        long evenSum = 0;                        // sum of all even numbers

        for (int num : arr) {
            if (Math.abs(num % 2) == 1) {        // Math.abs handles negative numbers
                oddSum  += num;                  // odd bucket
            } else {
                evenSum += num;                  // even bucket
            }
        }

        return new long[]{oddSum, evenSum};
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();                    // read array size
        int[] arr = new int[n];

        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();               // read each element
        }

        long[] result = oddEvenSum(arr);
        System.out.println(result[0]);           // odd sum
        System.out.println(result[1]);           // even sum

        sc.close();
    }
}